### PR TITLE
[RayJob] Enable job log streaming by setting `PYTHONUNBUFFERED` in job container

### LINKED ISSUE #1374: [Bug] RayJob does not stream logs in real time
### Search before asking

- [X] I searched the [issues](https://github.com/ray-project/kuberay/issues) and found no similar issues.

### KubeRay Component

ray-operator

### What happened + What you expected to happen

It is useful to inspect the logs of a long-running Kubernetes Job while the job is still running via `kubectl logs my-job-xxxxx` (the pod name) or `kubectl logs job/my-job`.  

However, this doesn't work with `RayJob`; the logs only show up once the entire job is finished.

Run the following `rayjob.yaml` with `kubectl apply -f rayjob.yaml`:

```yaml
apiVersion: ray.io/v1alpha1
kind: RayJob
metadata:
  name: rayjob-sample
spec:
  entrypoint: python /home/ray/samples/sample_code.py
  rayClusterSpec:
    rayVersion: '2.6.3' # should match the Ray version in the image of the containers
    # Ray head pod template
    headGroupSpec:
      # The `rayStartParams` are used to configure the `ray start` command.
      # See https://github.com/ray-project/kuberay/blob/master/docs/guidance/rayStartParams.md for the default settings of `rayStartParams` in KubeRay.
      # See https://docs.ray.io/en/latest/cluster/cli.html#ray-start for all available options in `rayStartParams`.
      rayStartParams:
        dashboard-host: '0.0.0.0'
      #pod template
      template:
        spec:
          containers:
            - name: ray-head
              image: rayproject/ray:2.6.3
              ports:
                - containerPort: 6379
                  name: gcs-server
                - containerPort: 8265 # Ray dashboard
                  name: dashboard
                - containerPort: 10001
                  name: client
              resources:
                limits:
                  cpu: "1"
                requests:
                  cpu: "200m"
              volumeMounts:
                - mountPath: /home/ray/samples
                  name: code-sample
          volumes:
            # You set volumes at the Pod level, then mount them into containers inside that Pod
            - name: code-sample
              configMap:
                # Provide the name of the ConfigMap you want to mount.
                name: ray-job-code-sample
                # An array of keys from the ConfigMap to create as files
                items:
                  - key: sample_code.py
                    path: sample_code.py

---
apiVersion: v1
kind: ConfigMap
metadata:
  name: ray-job-code-sample
data:
  sample_code.py: |
    import time
    for i in range(1, 121):
        print(i, flush=True)
        time.sleep(1)
```

Then run `watch kubectl logs job/rayjob-sample` (or find the pod with `kubectl get pod` and check the logs with `watch kubectl logs rayjob-sample-xxxx`).  **The logs only show up after 120s once the job is finished.**

Contrast this with an ordinary K8s job:

```yaml
apiVersion: batch/v1
kind: Job
metadata:
  name: number-printer-job
spec:
  template:
    spec:
      containers:
      - name: number-printer
        image: python:3.9-slim
        command:
        - python
        - -c
        - |
          import time
          for i in range(1, 121):
              print(i, flush=True)
              time.sleep(1)
      restartPolicy: Never
```

If you check `watch kubectl get logs job/number-printer-job`, you'll see that the logs are being streamed in real time.

### Reproduction script

One hint:

When running the k8s RayJob, the logs look like

```
(ray-py38) archit@archit-C02D27V8MD6N ray-operator % kubectl logs rayjob-sample-svx2z
2023-08-29 13:43:37,268 INFO cli.py:27 -- Job submission server address: http://rayjob-sample-raycluster-q9ndz-head-svc.default.svc.cluster.local:8265
2023-08-29 13:43:42,169 SUCC cli.py:33 -- ------------------------------------------------
2023-08-29 13:43:42,169 SUCC cli.py:34 -- Job 'rayjob-sample-2c6mj' submitted successfully
2023-08-29 13:43:42,169 SUCC cli.py:35 -- ------------------------------------------------
2023-08-29 13:43:42,169 INFO cli.py:240 -- Next steps
2023-08-29 13:43:42,169 INFO cli.py:241 -- Query the logs of the job:
2023-08-29 13:43:42,170 INFO cli.py:243 -- ray job logs rayjob-sample-2c6mj
2023-08-29 13:43:42,170 INFO cli.py:245 -- Query the status of the job:
2023-08-29 13:43:42,170 INFO cli.py:247 -- ray job status rayjob-sample-2c6mj
2023-08-29 13:43:42,170 INFO cli.py:249 -- Request the job to be stopped:
2023-08-29 13:43:42,170 INFO cli.py:251 -- ray job stop rayjob-sample-2c6mj
2023-08-29 13:43:42,174 INFO cli.py:258 -- Tailing logs until the job exits (disable with --no-wait):
1
2
3
``` 

whereas when running a ray job outside of K8s, it looks like

```
(ray-py38) archit@archit-C02D27V8MD6N ray-operator % ray job submit -- echo hi
Job submission server address: http://127.0.0.1:8265

-------------------------------------------------------
Job 'raysubmit_HZgmfaqeS5NiLmTN' submitted successfully
-------------------------------------------------------

Next steps
  Query the logs of the job:
    ray job logs raysubmit_HZgmfaqeS5NiLmTN
  Query the status of the job:
    ray job status raysubmit_HZgmfaqeS5NiLmTN
  Request the job to be stopped:
    ray job stop raysubmit_HZgmfaqeS5NiLmTN

Tailing logs until the job exits (disable with --no-wait):
hi

------------------------------------------
Job 'raysubmit_HZgmfaqeS5NiLmTN' succeeded
```

Note that there are no timestamps in this one.  I think this means some logging configuration is different in the two situations, but I'm not sure where it's configured.  Perhaps `PYTHONUNBUFFERED=1` needs to be set somewhere.

### Anything else

_No response_

### Are you willing to submit a PR?

- [X] Yes I am willing to submit a PR!

### PR DESCRIPTION
**Pull request by architkulkarni**
## Why are these changes needed?
Previously, `kubectl get logs job/my-rayjob` and `kubectl logs my-rayjob-xxxxx` (getting the logs from the pod directly) would be empty until the job finished.

This PR sets `PYTHONUNBUFFERED` in the job container so that the logs can be viewed while the job is still running.

This improves the log viewing experience to match that of ordinary Kubernetes Jobs.

This PR adds unit tests, but adding an end-to-end test (e.g. submit a job that prints a log and then sleeps forever, and check that we can read the log) will require some followup work. The main issue is the following:
- The RayJob sample YAML test framework waits for a job to "converge" (actually, to be `SUCCEEDED`) before running any Rules.  But for this test, we just want the job to start before running the Rule.

## Related issue number

Closes #1374 
## Checks

- [ ] I've made sure the tests are passing. 
- Testing Strategy
   - [x] Unit tests
   - [x] Manual tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: ray-operator/controllers/ray/rayjob_controller.go [hunk 1]
```diff
@@ -375,6 +376,26 @@ func (r *RayJobReconciler) getSubmitterTemplate(rayJobInstance *rayv1alpha1.RayJ
 		r.Log.Info("User-provided command is used", "command", submitterTemplate.Spec.Containers[0].Command)
 	}
 
+	// Set PYTHONUNBUFFERED=1 for real-time logging, unless user already set it.
```
Comment (kevin85421) [inline ray-operator/controllers/ray/rayjob_controller.go:379]: We can reuse `envVarExists`.

https://github.com/ray-project/kuberay/blob/aa1736359b7613a90767d3a79f3e436d1d593d52/ray-operator/controllers/ray/common/pod.go#L670

File: ray-operator/controllers/ray/rayjob_controller_unit_test.go [hunk 2]
```diff
@@ -134,4 +134,32 @@ func TestGetSubmitterTemplate(t *testing.T) {
 	assert.NoError(t, err)
 	assert.Equal(t, ([]string{"ray", "job", "submit", "--address", "http://test-url", "--", "echo", "hello", "world"}), submitterTemplate.Spec.Containers[0].Command)
 	assert.Equal(t, "rayproject/ray:custom-version", submitterTemplate.Spec.Containers[0].Image)
+
+	// Test 4: Check default PYTHONUNBUFFERED setting
+	submitterTemplate, err = r.getSubmitterTemplate(rayJobInstanceWithoutTemplate)
+	assert.NoError(t, err)
+	pythonUnbufferedSet := false
+	for _, envVar := range submitterTemplate.Spec.Containers[0].Env {
+		if envVar.Name == "PYTHONUNBUFFERED" {
```
Comment (kevin85421) [inline ray-operator/controllers/ray/rayjob_controller_unit_test.go:143]: Use `PythonUnbufferedEnvVarName` instead?

File: ray-operator/controllers/ray/rayjob_controller_unit_test.go [hunk 3]
```diff
@@ -134,4 +134,32 @@ func TestGetSubmitterTemplate(t *testing.T) {
...
+			assert.Equal(t, "1", envVar.Value)
+		}
+	}
+	assert.True(t, pythonUnbufferedSet)
+
+	// Test 5: Check user-provided PYTHONUNBUFFERED value is respected
+	userValue := "0"
+	rayJobInstanceWithTemplate.Spec.SubmitterPodTemplate.Spec.Containers[0].Env = []corev1.EnvVar{
+		{
+			Name:  "PYTHONUNBUFFERED",
```
Comment (kevin85421) [inline ray-operator/controllers/ray/rayjob_controller_unit_test.go:154]: Use `PythonUnbufferedEnvVarName` instead?

File: ray-operator/controllers/ray/rayjob_controller_unit_test.go [hunk 4]
```diff
@@ -134,4 +134,32 @@ func TestGetSubmitterTemplate(t *testing.T) {
...
+			Value: userValue,
+		},
+	}
+	submitterTemplate, err = r.getSubmitterTemplate(rayJobInstanceWithTemplate)
+	assert.NoError(t, err)
+	for _, envVar := range submitterTemplate.Spec.Containers[0].Env {
+		if envVar.Name == "PYTHONUNBUFFERED" {
```
Comment (kevin85421) [inline ray-operator/controllers/ray/rayjob_controller_unit_test.go:161]: Use `PythonUnbufferedEnvVarName` instead?

Comment (kevin85421) [inline ray-operator/controllers/ray/rayjob_controller_unit_test.go:161]: It is possible for the test to pass without entering this `if` statement.

Comment (architkulkarni) [reply ray-operator/controllers/ray/rayjob_controller_unit_test.go]: Good catch!

**Comment by architkulkarni** [comment]
@kevin85421 Comments have been addressed, please take another look!

File: ray-operator/controllers/ray/rayjob_controller.go [hunk 5]
```diff
@@ -375,6 +376,24 @@ func (r *RayJobReconciler) getSubmitterTemplate(rayJobInstance *rayv1alpha1.RayJ
...
+	userProvidedValue := ""
+	if common.EnvVarExists(PythonUnbufferedEnvVarName, submitterTemplate.Spec.Containers[0].Env) {
+		for _, envVar := range submitterTemplate.Spec.Containers[0].Env {
```
Comment (kevin85421) [inline ray-operator/controllers/ray/rayjob_controller.go:382]: It is a bit weird to iterate through the loop again after calling `EnvVarExists`.

Comment (kevin85421) [reply ray-operator/controllers/ray/rayjob_controller.go:382]: We may consider returning `bool, EnvVar`.

**Comment by architkulkarni** [comment]
@kevin85421 updated the PR to not check if the env var was already set, as discussed offline, because we don't anticipate any use case for `PYTHONUNBUFFERED=0`.

File: ray-operator/controllers/ray/rayjob_controller.go [hunk 6]
```diff
@@ -375,6 +376,12 @@ func (r *RayJobReconciler) getSubmitterTemplate(rayJobInstance *rayv1alpha1.RayJ
 		r.Log.Info("User-provided command is used", "command", submitterTemplate.Spec.Containers[0].Command)
 	}
 
+	// Set PYTHONUNBUFFERED=1 for real-time logging
+	submitterTemplate.Spec.Containers[0].Env = append(submitterTemplate.Spec.Containers[0].Env, v1.EnvVar{
```
Comment (kevin85421) [inline ray-operator/controllers/ray/rayjob_controller.go:380]: Nit: What will happen if users specify `PYTHONUNBUFFERED` on their own? I am not sure whether defining the same env variable twice in a container will throw an error message or override the first value.

**Review by kevin85421 — COMMENTED** [review]
Maybe we can use `submitterTemplate.Spec.Containers[common.RayContainerIndex]` to replace `submitterTemplate.Spec.Containers[0]`. You can either update it in the PR or open an issue as a good first issue.

**Review by kevin85421 — APPROVED** [review]

File: ray-operator/controllers/ray/rayjob_controller.go (hunk 6 above)
Comment (architkulkarni) [reply ray-operator/controllers/ray/rayjob_controller.go:380]: It seems the second definition will override the first one: https://www.baeldung.com/linux/kubernetes-pod-environment-variables

**Comment by kevin85421** [comment]
> Maybe we can use `submitterTemplate.Spec.Containers[common.RayContainerIndex]` to replace `submitterTemplate.Spec.Containers[0]`. You can either update it in the PR or open an issue as a good first issue.

Open a good first issue #1397.

**Pull request merged**